src/io_uring.c

```diff
@@ -0,0 +1,91 @@
+#include "io_uring.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+#include <string.h>
+
+#include "zmalloc.h"
+
+#define IO_URING_DEPTH 256 /* io_uring instance queue depth. */
+
+struct io_uring_context {
+    struct io_uring ring;
+    size_t queue_len;
+};
+
+io_uring_context *createIOUring(void) {
+    io_uring_context *uring_context = zmalloc(sizeof(struct io_uring_context));
+    uring_context->queue_len = 0;
+    struct io_uring_params params;
+    memset(&params, 0, sizeof(params));
+    /* On success, io_uring_queue_init_params(3) returns 0 and ring will
+     * point to the shared memory containing the io_uring queues.
+     * On failure -errno is returned. */
+    if (io_uring_queue_init_params(IO_URING_DEPTH, &uring_context->ring, &params) < 0) return NULL;
+    return uring_context;
+}
+
+/* Submit fdatasync request to io_uring. */
+int ioUringPrepFsyncAndSubmit(io_uring_context *ring_context, int fd) {
+    struct io_uring_sqe *sqe = io_uring_get_sqe(&ring_context->ring);
+    io_uring_prep_fsync(sqe, fd, IORING_FSYNC_DATASYNC);
+    ring_context->queue_len++;
+    int ret = io_uring_submit(&ring_context->ring);
+    if (ret < 0) {
+        errno = -ret;
+        return IO_URING_ERR;
+    }
+    return IO_URING_OK;
+}
+
+int ioUringWaitFsyncBarrier(io_uring_context *ring_context) {
```

Comment:
I like this design, but do you mind we implement it incrementally, maybe when we implement  https://github.com/valkey-io/valkey/pull/112? 

How about only keep one io-uring instance per server? 
```c
struct io_uring_context {
    struct io_uring ring;
    size_t fsync_queue_len;
    size_t write_queue_len;
};

int ioUringWaitWriteBarrier(io_uring_context *ring_context); /* Only consume write type requests from CQEs. */
```

Reply:
>How about only keep one io-uring instance per server?

I think we could and that will be fine with me too, though we will no longer need the new `io_uring_context`. 

>    size_t fsync_queue_len;
>    size_t write_queue_len;
    
If you'd like to use a single uring, I think we could combine the two counters too since anyways we can't pick which IO type to wait for and we will have to drain the CQ completely. I think we can tell the type by checking the user data associated with the CQE and handle the "post-processing" steps properly.


>maybe when we implement https://github.com/valkey-io/valkey/pull/112?

Generally I like the idea of "incremental perfection" and I have been reviewing this PR with the assumption that we can delay #112 a bit. But the more I review this PR, the more I feel that it would be better if we can think through io_uring holistically.